### PR TITLE
Fix multiple link insertion due to not unregistering the broadcast receiver, also provide a canonical path to the image editor

### PR DESCRIPTION
**Pull request by mehw**
Hello,

Apparently, there's a **bug** triggering via the `showInsertImageOrLinkDialog`, when an action is canceled (touching the **Cancel** or **Back** button), not unregistering an aborted action to the **broadcast receiver** may _accumulate_ it and **execute the next successful action a number of times more as much as there are registered actions**.

I'm asking, also, if it's possible to provide a **canonical path** to the image editor.  Writing `/path/to/file` and `/path/to/../to/file` give the same result, but the latter is a bit weird.  Sometimes, relative links may contain components, like `../`, than can be resolved.

Let me know,
Thanks.

# Multiple link insertion bug

1. Open a markdown file for editing;
2. Touch the **Insert Image** button;
3. Touch the **Browse Filesystem** button;
4. Touch the **Cancel** button;
5. Touch the **Gallery Picture** button;
6. Choose a picture to be inserted;
7. The link to the picture is inserted multiple times.

multiple_link_insertion

# Weird path provided to the image editor

Path components may be resolved to clear the location to the user.  In the example below, the problem is exaggerated.

weird_relative_path

### CODE REVIEW

**Review by gsantner — COMMENTED** [review]
This PR probably gets into conflict with <https://github.com/gsantner/markor/pull/2148/files>

**Comment by harshad1** [comment]
I have an alternate approach to this here: https://github.com/gsantner/markor/pull/2148

I will incorporate the usage of canonical path.

@mehw could you review my changes and make sure all the things / behaviours / bug-fixes from this PR also work as intended there?

If so we can just merge one.

File: app/src/main/java/net/gsantner/markor/frontend/AttachLinkOrFileDialog.java [hunk 1]
```diff
@@ -260,6 +261,11 @@ private static void insertItem(
         };
 
         final GsCallback.a1<File> insertFileLink = (file) -> {
+            // If path is empty, unregister the broadcast (autoUnregister == true)
+            if (file.toString().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (mehw) [reply app/src/main/java/net/gsantner/markor/frontend/AttachLinkOrFileDialog.java]: Thank you!

File: app/src/main/java/net/gsantner/markor/frontend/AttachLinkOrFileDialog.java [hunk 2]
```diff
@@ -302,20 +309,33 @@ private static void insertItem(
                 break;
             }
             case IMAGE_EDIT: {
-                if (pathEdit != null) {
+                if (pathEdit != null && !pathEdit.getText().toString().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (mehw) [reply app/src/main/java/net/gsantner/markor/frontend/AttachLinkOrFileDialog.java]: Thanks!

**Comment by mehw** [comment]
> This PR probably gets into conflict with https://github.com/gsantner/markor/pull/2148/files

@gsantner this was probably expected to happen https://github.com/gsantner/markor/pull/2144#issuecomment-1796653256, I based my work on the current logic, before #2148 was published...

> I have an alternate approach to this here: #2148
> 
> I will incorporate the usage of canonical path.
> 
> @mehw could you review my changes and make sure all the things / behaviours / bug-fixes from this PR also work as intended there?
> 
> If so we can just merge one.

@harshad1 I'll take a look, I'm rebasing on master in the meantime.

**Comment by harshad1** [comment]
My PR is ready to merge too. It uses a weakref / callback approach for greater control.

Please let me know if there are any changes (from this PR or otherwise) which you'd like to incorporate into mine.

**Comment by gsantner** [comment]
Due other PR's edited the same files/things, there are merge conflicts now.

Can you please fix it?

**Comment by harshad1** [comment]
@mehw could you check if all the bugs you're addressing here have been fixed in master now.

**Comment by mehw** [comment]
@harshad1 @gsantner I'm sorry fellow developers, I was "_out of commission_" for quite some time... I'll try to catch up in these days.

**Comment by harshad1** [comment]
Closing as I have a fully working alternative using callbacks already merged into master. Please let me know if this should be reopened for whatever reason.

https://github.com/gsantner/markor/blob/master/app/src/main/java/net/gsantner/opoc/util/GsContextUtils.java#L1719

**Comment by mehw** [comment]
> Closing as I have a fully working alternative using callbacks already merged into master. Please let me know if this should be reopened for whatever reason.
> 
> https://github.com/gsantner/markor/blob/master/app/src/main/java/net/gsantner/opoc/util/GsContextUtils.java#L1719

Thanks ;)

**Pull request closed without merging**